File: webreport/frontend-web/src/answer-visuals.ts
/** Versioned, data-only blocks embedded in an assistant's Markdown response. */
export const VISUAL_LANGUAGE = "dig-visual";
export const MAX_VISUAL_LENGTH = 64_000;
export const MAX_VISUAL_BLOCKS = 4;

type MarkdownNode = {
  type: string;
  lang?: string | null;
  children?: MarkdownNode[];
};
/** Each Markdown tree gets its own budget, including blocks nested in lists/quotes. */
export function limitVisualBlocks() {
  return (tree: MarkdownNode) => {
    let count = 0;
    const visit = (node: MarkdownNode) => {
      if (
        node.type === "code" &&
        node.lang === VISUAL_LANGUAGE &&
        ++count > MAX_VISUAL_BLOCKS
      )
        node.lang = `${VISUAL_LANGUAGE}-limit`;
      node.children?.forEach(visit);
    };
    visit(tree);
  };
}
export type ValueFormat = "number" | "percent" | "duration";
type Presentation = { format: ValueFormat; unit?: string };
type BaseVisual = {
  version: 1;
  title: string;
  subtitle?: string;
  source: string;
  note?: string;
};
export type MetricCard = Presentation & {
  key: string;
  label: string;
  value: number | null;
  change_percent?: number;
  comparison?: string;
  sentiment: "good" | "bad" | "neutral";
  trend?: (number | null)[];
  trend_label?: string;
};
export type MetricsVisual = BaseVisual & {
  type: "metrics";
  items: MetricCard[];
};
export type ChartVisual = BaseVisual &
  Presentation & {
    type: "chart";
    kind: "line" | "area" | "bar";
    x_key: string;
    x_label?: string;
    series: { key: string; label: string }[];
    data: Record<string, string | number | null>[];
  };
export type AnswerVisual = MetricsVisual | ChartVisual;

const object = (value: unknown): Record<string, unknown> => {
  if (!value || typeof value !== "object" || Array.isArray(value))
    throw new Error("Expected an object");
  return value as Record<string, unknown>;
};
const text = (value: unknown, max = 120): string => {
  if (typeof value !== "string" || !value.trim() || value.length > max)
    throw new Error("Invalid label");
  return value.trim();
};
const optionalText = (value: unknown, max: number) =>
  value === undefined ? undefined : text(value, max);
const numeric = (value: unknown): number => {
  if (
    typeof value !== "number" ||
    !Number.isFinite(value) ||
    Math.abs(value) > 1e15
  )
    throw new Error("Invalid number");
  return value;
};
const nullableNumber = (value: unknown) =>
  value === null ? null : numeric(value);
const key = (value: unknown): string => {
  const result = text(value, 48);
  if (
    !/^[a-z][a-z0-9_]*$/.test(result) ||
    ["constructor", "prototype"].includes(result)
  )
    throw new Error("Invalid data key");
  return result;
};
function list(value: unknown, max: number): unknown[] {
  if (!Array.isArray(value) || !value.length || value.length > max)
    throw new Error("Invalid item count");
  return value;
}
function unique(keys: string[]) {
  if (new Set(keys).size !== keys.length) throw new Error("Duplicate key");
}
function presentation(raw: Record<string, unknown>): Presentation {
  const format = raw.format ?? "number";
  if (format !== "number" && format !== "percent" && format !== "duration")
    throw new Error("Unknown number format");
  const unit = optionalText(raw.unit, 16);
  if (unit && format !== "number") throw new Error("Conflicting units");
  return { format, unit };
}

/** Project known fields only. Never evaluate code, resolve URLs or pass model props to a chart. */
export function parseVisual(source: string): AnswerVisual | null {
  try {
    if (source.length > MAX_VISUAL_LENGTH) return null;
    const raw = object(JSON.parse(source));
    if (raw.version !== 1) return null;
    const base: BaseVisual = {
      version: 1,
      title: text(raw.title),
      subtitle: optionalText(raw.subtitle, 240),
      source: text(raw.source, 240),
      note: optionalText(raw.note, 500),
    };
    if (raw.type === "metrics") {
      const items = list(raw.items, 6).map((item): MetricCard => {
        const card = object(item);
        const change =
          card.change_percent === undefined
            ? undefined
            : numeric(card.change_percent);
        const comparison = optionalText(card.comparison, 120);
        if ((change !== undefined) !== (comparison !== undefined))
          throw new Error("A change needs a comparison period");
        const sentiment = card.sentiment ?? "neutral";
        if (!["good", "bad", "neutral"].includes(String(sentiment)))
          throw new Error("Unknown change meaning");
        const trend =
          card.trend === undefined
            ? undefined
            : list(card.trend, 60).map(nullableNumber);
        const trendLabel = optionalText(card.trend_label, 120);
        if (
          trend &&
          (!trendLabel || trend.length < 2 || trend.every((v) => v === null))
        )
          throw new Error("A trend needs observed values and a period");
        return {
          key: key(card.key),
          label: text(card.label, 80),
          value: nullableNumber(card.value),
          ...presentation(card),
          change_percent: change,
          comparison,
          sentiment: sentiment as MetricCard["sentiment"],
          trend,
          trend_label: trendLabel,
        };
      });
      unique(items.map((item) => item.key));
      return { ...base, type: "metrics", items };
    }
    if (
      raw.type !== "chart" ||
      !["line", "area", "bar"].includes(String(raw.kind))
    )
      return null;
    const xKey = key(raw.x_key);
    const series = list(raw.series, 4).map((item) => {
      const spec = object(item);
      return { key: key(spec.key), label: text(spec.label, 80) };
    });
    unique([xKey, ...series.map((s) => s.key)]);
    const data = list(raw.data, 120).map((item) => {
      const row = object(item);
      const result: Record<string, string | number | null> = {
        [xKey]: text(row[xKey], 100),
      };
      for (const s of series) result[s.key] = nullableNumber(row[s.key]);
      return result;
    });
    unique(data.map((row) => String(row[xKey])));
    if (series.some((s) => data.every((row) => row[s.key] === null)))
      return null;
    return {
      ...base,
      ...presentation(raw),
      type: "chart",
      kind: raw.kind as ChartVisual["kind"],
      x_key: xKey,
      x_label: optionalText(raw.x_label, 80),
      series,
      data,
    };
  } catch {
    return null;
  }
}

// Stable colors depend on a metric key, never its position in a particular answer.
const palette = [
  "#8061ad",
  "#cc655b",
  "#3c867c",
  "#ab7b30",
  "#557dae",
  "#a25d83",
];
const metricColors: Record<string, string> = {
  users: palette[0],
  visits: palette[1],
  pageviews: palette[2],
  revenue: palette[3],
  conversions: palette[4],
  conversion_rate: palette[4],
  bounce_rate: palette[5],
};
export function visualColor(metric: string): string {
  if (Object.hasOwn(metricColors, metric)) return metricColors[metric];
  let hash = 0;
  for (const char of metric)
    hash = (Math.imul(hash, 31) + char.charCodeAt(0)) | 0;
  return palette[(hash >>> 0) % palette.length];
}
export function visualNumber(
  value: number | null,
  format: ValueFormat = "number",
  unit?: string,
  compact = false,
): string {
  if (value === null) return "—";
  if (format === "duration") {
    const seconds = Math.round(Math.abs(value));
    const hours = Math.floor(seconds / 3600);
    const minutes = Math.floor((seconds % 3600) / 60);
    return `${value < 0 ? "−" : ""}${hours ? `${hours} ч ` : ""}${minutes ? `${minutes} мин ` : ""}${seconds % 60 || (!hours && !minutes) ? `${seconds % 60} с` : ""}`.trim();
  }
  const number = new Intl.NumberFormat("ru-RU", {
    maximumFractionDigits: 6,
    ...(value !== 0 && Math.abs(value) < 1e-6
      ? { notation: "scientific" as const }
      : {}),
    ...(compact
      ? { notation: "compact" as const, maximumFractionDigits: 1 }
      : {}),
  }).format(value);
  return number + (format === "percent" ? "%" : unit ? ` ${unit}` : "");
}

/** Copy a readable text equivalent instead of implementation JSON. */
export function copyAnswer(text: string): string {
  return text.replace(
    /^(`{3,}|~{3,})dig-visual[^\S\n]*\r?\n([\s\S]*?)^\1[^\S\n]*$/gm,
    (original, _fence, source) => {
      const block = parseVisual(source);
      if (!block) return original;
      const escape = (value: string) =>
        value.replace(/\|/g, "\\|").replace(/\r?\n/g, " ");
      const lines = [block.title, block.subtitle || ""];
      if (block.type === "metrics") {
        lines.push(
          ...block.items.map(
            (item) =>
              `${item.label}: ${visualNumber(item.value, item.format, item.unit)}` +
              (item.change_percent !== undefined
                ? ` (${item.change_percent > 0 ? "+" : ""}${visualNumber(item.change_percent, "percent")}; ${item.comparison})`
                : ""),
          ),
        );
      } else {
        lines.push(
          [
            `| ${[block.x_label || block.x_key, ...block.series.map((s) => s.label)].map(escape).join(" | ")} |`,
            `| ${["---", ...block.series.map(() => "---:")].join(" | ")} |`,
            ...block.data.map(
              (row) =>
                `| ${[String(row[block.x_key]), ...block.series.map((s) => visualNumber(row[s.key] as number | null, block.format, block.unit))].map(escape).join(" | ")} |`,
            ),
          ].join("\n"),
        );
      }
      lines.push(block.source, block.note || "");
      return lines.filter(Boolean).join("\n\n");
    },
  );
}

File: webreport/backend/workspace/answer_visuals.py
"""Formatting instructions for the React chat; visual blocks travel in message.content."""

import json
from pathlib import Path

# Synthetic examples are also parsed by the frontend contract tests.
VISUAL_EXAMPLES = json.loads(
    Path(__file__).with_name("answer_visual_examples.json").read_text(encoding="utf-8")
)
VISUAL_PROMPT = """

## Visual answers in this chat
The chat renders version 1 JSON objects inside fenced Markdown blocks whose language
is dig-visual. Place a block on its own between paragraphs exactly where it is useful.
Use cards for a summary of key metrics; use a line/area chart for time trends or a bar
chart for category comparisons. Prefer these visuals for requests about dynamics,
comparisons or metric summaries, unless the user requests text only. Keep ordinary
questions concise. Explain the main finding in prose; do not just emit a visual.
Use at most 4 visual blocks per answer. Do not wrap the whole answer in a code fence.
The application owns styling, colors and interactions inside these blocks. Inside a
dig-visual block emit data only: no HTML, JavaScript, CSS, image URLs or chart-library
options. Other tools' file attachments remain separate from this inline visual format.

Every block requires version:1, type, title (up to 120 characters), and source (up to
240 characters, an accurate human-readable data source). Optional subtitle (240)
should identify period and granularity; optional note (500) must disclose sampling,
limits, missing data or other caveats when applicable. Answer labels in the user's language.
Only use values you actually obtained through tools or explicit values the user gave;
computed metrics must follow explained arithmetic. Example values below are synthetic,
never facts about the user's project. If values are unavailable, say so; do not draw
plausible-looking trends, fabricate comparison percentages or substitute zero for unknown.
Read the needed rows before building a visual and still follow the mark_report rule
for analytics handles. A visual is presentation, not another data source or a tool call.

type:"metrics": items is an array of 1–6 objects, each with key, label (80), value
(a JSON number or null), and optionally format (number/percent/duration; default number),
unit (up to 16 characters, only with format:number), change_percent (JSON number),
comparison (up to 120 characters describing the real baseline; REQUIRED with change_percent),
sentiment (good/bad/neutral, default neutral). Never assume an increase is beneficial;
use good/bad only when the project goal supports this interpretation. Optional trend
contains 2–60 observed numeric or null values in chronological order and REQUIRES
trend_label (up to 120 characters identifying its period). Do not sum daily unique
users to obtain period uniques. Do not invent a trend to decorate a card.

type:"chart": kind is line/area/bar; x_key identifies the category/date field;
optional x_label names it. series contains 1–4 {key,label} objects and data contains
1–120 row objects. Each row must contain a UNIQUE nonempty string category under x_key
(up to 100 characters) and a JSON number or explicit null for EVERY series key.
Every series must have at least one observed numeric value. Dates use ISO format and
chronological order; include missing expected time buckets as null so gaps stay visible.
All series share one vertical axis: compare only compatible units. Optional block-level
format and unit follow the same rules as cards. Percent values are in percentage units
(3.8 means 3.8%, not 380%); duration values are seconds. Bar charts start at zero.
If there are too many rows, request suitable aggregation or clearly label a top-N subset;
do not silently truncate a trend. Use a Markdown table when a visual would mislead.

All keys must start with a lowercase ASCII letter and contain only lowercase letters,
digits or underscores, at most 48 characters; no dots, brackets, constructor or prototype.
Keys within a block must be distinct, including x_key. Keep the SAME metric key across
answers and blocks so colors stay consistent. Prefer users, visits, pageviews, revenue,
conversions, conversion_rate and bounce_rate when those metrics actually match the data.
Numbers must be finite JSON numbers with absolute value at most 1e15. A block must fit
within 64,000 characters. Never place number strings with spaces or % inside numeric fields.

These two examples show the exact syntax; replace all synthetic values and labels:
""" + "\n\n".join(
    "```dig-visual\n" + json.dumps(example, ensure_ascii=False, separators=(",", ":")) + "\n```"
    for example in VISUAL_EXAMPLES
)
